framework/src/materials/ParsedMaterialHelper.C

```diff
@@ -176,26 +181,37 @@ ParsedMaterialHelper::functionsOptimize()
     mooseInfo("Failed to JIT compile expression, falling back to byte code interpretation.");
 }
 
+template <>
 void
-ParsedMaterialHelper::initQpStatefulProperties()
+ParsedMaterialHelper<true>::functionsOptimize()
+{
+  // base function
+  if (!_disable_fpoptimizer)
+    _func_F->Optimize();
+  if (!_enable_jit || !_func_F->JITCompile())
+    mooseError("ADParsedMaterials require JIT compilation to be enabled and working.");
+}
+
+template <bool is_ad>
+void
+ParsedMaterialHelper<is_ad>::initQpStatefulProperties()
 {
   if (_prop_F)
     (*_prop_F)[_qp] = 0.0;
 }
 
+template <bool is_ad>
 void
-ParsedMaterialHelper::computeQpProperties()
+ParsedMaterialHelper<is_ad>::computeQpProperties()
 {
-  Real a;
-
   // fill the parameter vector, apply tolerances
   for (unsigned int i = 0; i < _nargs; ++i)
   {
     if (_tol[i] < 0.0)
       _func_params[i] = (*_args[i])[_qp];
     else
     {
-      a = (*_args[i])[_qp];
+      auto a = (*_args[i])[_qp];
```

Comment:
I'm always curious whether to optimize for the probably much more popular use case which is builtins vs. optimize for the potentially much more expensive but maybe much less used user-defined type. I'll let you decide whether you want to copy here or catch as `const auto &`

Reply:
Since FParser doesn't know `MooseArray`s I have to prepare a separate parameter vector. Good point about the copying though. I could add a separate mode in which parameters are passed as a vector of pointers rather than a vector of copied values. I would prefer to leave that for the next steps though.